<H1><center>Project 4<BR><BR>
Recommender Systems</center></H1>

<H2>Task 1: Algorithmic Bias
</H2>

<P>Read <a href="https://cacm.acm.org/magazines/2016/10/207759-battling-algorithmic-bias/abstract">this article</a> (also available <a href="https://drive.google.com/file/d/1bRKH--BoCAQw7ZyK8xmBGGZR2lyPlRaN/view?usp=sharing">here</a>) about bias in algorithms.</P>

<P>Please address (minimum 200 words) the following questions in the space below. As the article describes, "common wisdom among programmers is to develop a pure algorithm that does not incorporate protected attributes into the model," where protected classes may include aspects such as race, gender, age, sexual orientation, and disability status. As a result, the algorithms can be bias. How might algorithms be evaluated for bias? Are there machine learning applications where you would sacrifice accuracy in order to reduce bias? Are there machine learning applications where you would allow bias in order to increase accuracy?</P>

One of themost important thing a programmer can do is to work to reduce bias. If humans have bias and it inherently affects our dicision making and we have a way of mitigating that bias within our algoirthms which we also use for decision making, that should be our prioity. Obviously there is a line between compriamising the accuracy of the model in order to make it less biased, but overall I do think reducing bias is more important and if the accuracy decreases a little bit in my opinion that is fine. I have actually taught a course on the Northpointe algorithm and it has terrible rates of false identification for Black people, and it is impossible for computer scientists to be able to fix the issue because it is a black box algorithm. I think one of the most obvious solutions to making sure an algorithm can be evaluated for bias is training it for different data sets. One data set might include more diversely ranging data while the other could be more narrow, and depending on how the algorithm performs on the two different sets of data, some idea of how biased the algorithm is could be gathered. Other ways to measure bias could include seeing how data sets for minorities compare to the majority and seeing if the algorithm results in different predictions. I absolutely think there would be machine learning models where I would sacrifice accuracy to reduce bias. Northopinte being a prime example of this. I think it is more important to make sure populations that are already harmed by systems in place be protected my algoirthms whenever possible. I can't think of machine learning applications where I wouldn't want to reduce bias, maybe if the data was related to a biological, mathematical or checmical study then sure. But if the data concerns human lives and well-being I think reducing bias is of the utmost importance. 








<H2>Task 2: Pipeline for Classification
</H2>

<P>As broader context for this project, little or no starter code is provided for the tasks in the project. One of the goals of this lack of starter code is to give you an opportunity to practice developing meaningful computational artifacts mostly or entirely from scratch.</P>

<P>To start, let's make a simple pipeline that integrates in one function some of the aspects of supervised classification.</P>

<P>Create a function named <code>pipeline</code> that takes three arguments:</P>
<ol>
    <li>The name of a comma separated values (CSV) file. Assume the file has a header row and all subsequent rows contain numerical data. The last column in the file corresponds to classification labels.</li>
    <li>A Boolean indicating whether feature scaling should be performed on the data in the file.</li>
    <li>A supervised classification model, such as a random forest classifier, a <em>k</em> nearest neighbors classifier, a logistic regression classifier, or a support vector machine.</li>
</ol>

<P>Your <code>pipeline</code> function should:</P>
<ol>
    <li>read in the specified file</li>
    <li>extract the feature vectors (e.g., <em>X</em>) and labels (e.g., <em>y</em>)</li>
    <li>split the data into training (80%) and testing (20%) (with <code>random_state=0</code>)</li>
    <li>perform feature scaling <em>if</em> the specified Boolean argument is <code>True</code></li>
    <li>train the classification model on the training data</li>
    <li>print out the accuracy of the classification model on the testing data</li>
</ol>


In [1]:
# The *pipeline* function executes some of the aspects of supervised
# classification on a set of data.
# The function takes three arguments: the name of a CSV file with a header line and
# whose last column contains labels, a Boolean indicating if feature scaling
# should be performed on the data in the file, and a supervised classification model.
# The function reads in data from the file, splits the data into training (80%)
# and testing (20%), performs feature scaling if specified by the Boolean argument,
# trains the model, and prints out the model's accuracy on the testing data.
import numpy as np
import pandas as pd

from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import precision_score




def pipeline(csv_file, feature_scaling, classification):
    DATA = np.loadtxt(csv_file, delimiter=',', skiprows=1)  # Read data from comma delimited file
    X = DATA[:, :-1]
    y = DATA[:, -1] # get the last column of the rows (should be 0 or 1)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=0)
    scaler = StandardScaler()
    X_train_scaled = scaler.fit(X_train, y_train)
    X_train_scaled_transform = scaler.transform(X_train)
    X_test_transform = scaler.transform(X_test)
    if feature_scaling == True and classification == 'knn': 
        knn = KNeighborsClassifier()
        knn.fit(X_train_scaled_transform, y_train)
        y_pred= knn.predict(X_test_transform)
        pscore = precision_score(y_test, y_pred)
        return(pscore)
    elif feature_scaling == True and classification == 'logistic':
        logistic = LogisticRegression(random_state=0)
        logistic.fit(X_train_scaled_transform, y_train)
        y_pred= logistic.predict(X_test_transform)
        pscore = precision_score(y_test, y_pred)
        return(pscore)
    elif feature_scaling == True and classification == 'random_forest':
        random_forest = RandomForestClassifier(random_state=0)
        random_forest.fit(X_train_scaled_transform, y_train)
        y_pred= random_forest.predict(X_test_transform)
        pscore = precision_score(y_test, y_pred)
        return(pscore)
    elif feature_scaling == True and classification == 'svm':
        svc = SVC(random_state= 0)
        svc.fit(X_train_scaled_transform, y_train)
        y_pred= svc.predict(X_test_transform)
        pscore = precision_score(y_test, y_pred)
        return(pscore)
    elif feature_scaling == False and classification == 'logistic':
        logistic = LogisticRegression(random_state=0)
        logistic.fit(X_train, y_train)
        test_score = logistic.score(X_test, y_test)
        return(test_score)
    elif feature_scaling == False and classification == 'random_forest':
        random_forest = RandomForestClassifier(random_state=0)
        random_forest.fit(X_train, y_train)
        test_score = random_forest.score(X_test, y_test)
        return(test_score)
    elif feature_scaling == False and classification == 'knn':    
        knn = KNeighborsClassifier()
        knn.fit(X_train, y_train)
        test_score = knn.score(X_test, y_test)
        return(test_score)
    elif feature_scaling == False and classification == 'svm':    
        svc = SVC(random_state= 0)
        svc.fit(X_train, y_train)
        test_score = svc.score(X_test, y_test)
        return(test_score)









In [2]:
pipeline('admission.csv', False, 'random_forest')

1.0

<P>With the <code>pipeline</code> function above, you have a mechanism for quickly invoking several aspects of supervised classification. Below, you should experiment with executing your <code>pipeline</code> function on the following four datasets:</P>
<ol>
    <li>The file <code>admission.csv</code> contains GPA and GMAT score data on people who have/haven't been <a href="https://www.kaggle.com/willianleite/mba-admission">admitted to an MBA program</a> (graduate school in business). How well can you predict who will be admitted based on their GPA and their GMAT standardized test score?</li>
    <li>The file <code>dementia.csv</code> contains MRI scans and other health data on people who do/don't have <a href="https://www.kaggle.com/shashwatwork/dementia-prediction-dataset">dementia</a>. How well can you predict who has dementia based on this health data?</li>
    <li>The file <code>star.csv</code> contains chromaticity and spectral data for different stars in the galaxy. In this <a href="https://www.kaggle.com/vinesmsuic/star-categorization-giants-and-dwarfs">stellar classification problem</a>, how well can you determine which stars are "giants" and which are "dwarfs".</li>
    <li>The file <code>airlineDelays.csv</code> contains data on airplane flights and whether or not the flight was delayed. Data include information on the airline, the departing airport of the flight, the weather, the number of seats on the flight, the number of flight attendants, the age of the plane, etc. How well can you predict <a href="https://www.kaggle.com/threnjen/2019-airline-delays-and-cancellations">flight delays</a>?</li>
</ol>

In [3]:
# Experiments executing *pipeline* on four different data sets

print(pipeline('admission.csv', False, 'random_forest'))
print(pipeline('dementia.csv', True, 'svm'))
print(pipeline('star.csv', True, 'knn'))
print(pipeline('airlineDelays.csv', True, 'logistic'))

1.0
0.9722222222222222
0.8528557599225557
0.6253902185223725


<P><font color="maroon"><u>For the <code>admission</code> data, what is the testing accuracy <code>without</code> feature scaling and using a <code>Random Forest</code> classifier (with <code>random_state=0</code>)?<u></font></P>

100%

<P><font color="maroon"><u>For the <code>dementia</code> data, what is the testing accuracy <code>with</code> feature scaling and using a <code>Support Vector Machine</code> (with <code>random_state=0</code>)?<u></font></P>

97.22%

<P><font color="maroon"><u>For the <code>star</code> data, what is the testing accuracy <code>with</code> feature scaling and using a <code><em>k</em> Nearest Neighbors</code> classifier?<u></font></P>

85.29%

<P><font color="maroon"><u>For the <code>airline delays</code> data, what is the testing accuracy <code>with</code> feature scaling and using a <code>Logistic Regression</code> classifier (with <code>random_state=0</code>)?<u></font></P>

62.54%

<H2>Task 3: Sentiment Analysis of Movie Reviews
</H2>

<P>Remember in Exercise 3 where you performed sentiment analysis of Twitter data? If not, reviewing it will be very helpful here :). Here you will perform sentiment analysis of <a href="https://www.kaggle.com/lakshmi25npathi/imdb-dataset-of-50k-movie-reviews">movie reviews</a>, i.e., you will determine whether a movie review is negative or positive based on the text of the review.</P>
<P>The file <code>movie_reviews.txt</code> contains 50,000 lines. Each line contains a review of a movie followed by a label, either "@@@negative" or "@@@positive", indicating if the review is negative or positive. Your task is to:</P>
<ol>
    <li>read in the data from the file</li>
    <li>randomly shuffle the data, while making sure each label stays with its corresponding review</li>
    <li>store the data as a corpus and as labels</li>
    <li>split the data into training (80%) and testing (20%)</li>
    <li>use the Bag of Words approach (tf-idf) to vectorize the data (you should use unigrams rather than bigrams here)</li>
    <li>use a logistic regression classifier to predict the sentiment of reviews</li>
    <li>based on the logistic regression model weights, determine the 10 lowest weighted words (indicative of negative sentiment reviews) and the 10 highest weighted words (indicative of positive sentiment reviews)</li>
</ol>

In [52]:
# Read in movie review data from file.
# Randomly shuffle data, making sure each label stays with its corresponding review.
# Store data as corpus and labels.

import random
random.seed(42)
def readFile(filename):
    DATA = []
    in_file = open(filename, 'r')
    line = in_file.readline()  # Ignore header line
    line = in_file.readline()
    while (line != ''):
        try: 
            if (line.strip() == ''): 
                line = in_file.readline()
                continue
            firstCommaIndex = line.rfind('@@@')
            label = line[(firstCommaIndex+3):]
            review = line[:(firstCommaIndex)]
            DATA.append([review, label])
            line = in_file.readline()
        except (UnicodeDecodeError, ValueError): line = ' '
    in_file.close()
    random.shuffle(DATA)  # Shuffle
    return [row[0] for row in DATA], [row[1] for row in DATA]

corpus, labels = readFile('movie_reviews.txt')







In [53]:
# Separate into training and testing data

TEST_SIZE = 0.2  # 20% testing data and 80% training data

separator = int((1.0 - TEST_SIZE)*len(corpus))
corpus_train = corpus[:separator]
labels_train = labels[:separator]
corpus_test = corpus[separator:]
labels_test = labels[separator:]
print(len(corpus_train))
print(len(labels_train))
print(len(corpus_test))
print(len(labels_test))





39999
39999
10000
10000


In [54]:
# Text feature extraction (using Bag of Words approach with unigrams) for training data

# Text feature extraction for training data
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer()  # Use individual words as tokens
X_train = vectorizer.fit_transform(corpus_train)
y_train = np.array(labels_train)
print(X_train.shape)
print(y_train.shape)





(39999, 92886)
(39999,)


In [55]:
# Text feature extraction (using Bag of Words approach with unigrams) for testing data

X_test = vectorizer.transform(corpus_test)
y_test = np.array(labels_test)
print(X_test.shape)
print(y_test.shape)


(10000, 92886)
(10000,)


In [56]:
# Use logistic regression to predict sentiment of reviews (with random_state=0)

logistic = LogisticRegression(random_state=0)
logistic.fit(X_train, y_train)
test_score = logistic.score(X_test, y_test)
print(test_score)



0.898


/Users/zeynoyalcin/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [57]:
# Get Logistic Regression weights
learner = LogisticRegression(random_state=0)
learner.fit(X_train, y_train)
weights = learner.coef_  # Get the learned Perceptron weights
sorted_weights = np.argsort(weights)  # Sort
features = vectorizer.get_feature_names()  # Get the features
print('\nLowest weighted words (indicative of negative sentiment tweets)')
for i in range(25): print(features[sorted_weights[0,i]])
print('\nHighest weighted words (indicative of positive sentiment tweets)')
for i in range(25): print(features[sorted_weights[0,len(sorted_weights[0])-i-1]])






Lowest weighted words (indicative of negative sentiment tweets)
worst
bad
awful
waste
boring
terrible
poor
nothing
worse
horrible
dull
poorly
stupid
minutes
fails
disappointing
annoying
disappointment
unfortunately
supposed
no
lame
script
ridiculous
save

Highest weighted words (indicative of positive sentiment tweets)
great
excellent
best
perfect
amazing
wonderful
loved
enjoyed
brilliant
today
favorite
hilarious
fun
well
definitely
love
enjoyable
highly
superb
it
fantastic
entertaining
surprised
still
dvd


/Users/zeynoyalcin/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


<P><font color="maroon"><u>What is the testing accuracy of your logistic regression model at determining the sentiment of movie reviews?<u></font></P>

89.8% accurate 

<P><font color="maroon"><u>For your logistic regression classifier, what are the 10 lowest weighted words, indicative of negative reviews?<u></font></P>

worst,
bad,
awful,
waste,
boring,
terrible,
poor,
nothing, 
worse,
horrible

<P><font color="maroon"><u>For your logistic regression classifier, what are the 10 highest weighted words, indicative of positive reviews?<u></font></P>

great,
excellent,
best,
perfect,
amazing,
wonderful,
loved,
enjoyed,
brilliant,
today

<H2>Task 4: Detecting Fake News
</H2>

<P>Can you use machine learning to distinguish <a href="https://www.kaggle.com/clmentbisaillon/fake-and-real-news-dataset">fake news stories from true news stories</a>? You are provided two files, each containing over 20,000 lines, where each line corresponds to a news story. The file <code>Fake_News.txt</code> contains fake news stories. The file <code>True_News.txt</code> contains true news stories. Using the Bag of Words (tf-idf) approach, your task is to distinguish fake news stories from true news stories. We are not providing much guidance here (and no starter code) so that you can think about how best to tackle this problem largely from scratch.</P>

In [148]:
import random
random.seed(42)
def readFile(filename, filename1):
    DATA = []
    in_file = open(filename, 'r')
    in_file1 = open(filename1, 'r')

    line = in_file.readline()  # Ignore header line
    line1 = in_file1.readline()
    while (line != ''):
        try: 
            if (line.strip() == ''): 
                line = in_file.readline()
                continue
            firstCommaIndex = line.rfind('/n')
            label = 0
            news = line[:(firstCommaIndex)]
            DATA.append([news, label])
            line = in_file.readline()
            if (line1.strip() == ''): 
                line1 = in_file1.readline()
                continue
            firstCommaIndex = line1.rfind('/n')
            label = 1 
            news = line1[:(firstCommaIndex)]
            DATA.append([news, label])
            line1 = in_file1.readline()
        except (UnicodeDecodeError, ValueError): line = ' '
    in_file.close()
    in_file1.close()
    random.shuffle(DATA)  # Shuffle
    return [row[0] for row in DATA], [row[1] for row in DATA]

corpus, labels = readFile('Fake_News.txt', 'True_News.txt')


In [149]:
# Your code here. 
# Rather than put all your code in this one cell, create multiple new cells as needed.

TEST_SIZE = 0.2  # 20% testing data and 80% training data

separator = int((1.0 - TEST_SIZE)*len(corpus))
corpus_train = corpus[:separator]
labels_train = labels[:separator]
corpus_test = corpus[separator:]
labels_test = labels[separator:]
print(len(corpus_train))
print(len(labels_train))
print(len(corpus_test))
print(len(labels_test))







35405
35405
8852
8852


In [150]:
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer()  # Use individual words as tokens
X_train = vectorizer.fit_transform(corpus_train)
y_train = np.array(labels_train)
print(X_train.shape)
print(y_train.shape)

(35405, 110408)
(35405,)


In [151]:
X_test = vectorizer.transform(corpus_test)
y_test = np.array(labels_test)
print(X_test.shape)
print(y_test.shape)


(8852, 110408)
(8852,)


In [152]:
logistic = LogisticRegression(random_state=0)
logistic.fit(X_train, y_train)
test_score = logistic.score(X_test, y_test)
print(test_score)

0.9783099864437416


In [153]:
learner = LogisticRegression(random_state=0)
learner.fit(X_train, y_train)
weights = learner.coef_  # Get the learned Perceptron weights
sorted_weights = np.argsort(weights)  # Sort
features = vectorizer.get_feature_names()  # Get the features
print('\nLowest weighted words (indicative of Fake news)')
for i in range(25): print(features[sorted_weights[0,i]])
print('\nHighest weighted words (indicative of True news)')
for i in range(25): print(features[sorted_weights[0,len(sorted_weights[0])-i-1]])


Lowest weighted words (indicative of Fake news)
via
this
us
read
com
just
gop
that
image
mr
featured
is
hillary
you
watch
obama
pic
america
here
https
sen
even
these
getty
rep

Highest weighted words (indicative of True news)
said
on
in
republican
told
presidential
minister
reporters
had
nov
democratic
spokesman
edt
monday
comment
its
representatives
statement
saying
government
london
election
united
senate
ly


<P><font color="maroon"><u>What is the testing accuracy of your machine learning method in determining whether a news story is fake?<u></font></P>

97.8% accurate 

<P><font color="maroon"><u>What are 10 words highly associated with true news stories?<u></font></P>

said,
on,
in,
republican,
told,
presidential,
minister,
reporters,
had,
nov,
democratic

<P><font color="maroon"><u>What are 10 words highly associated with fake news stories?<u></font></P>

via,
this,
us,
read,
com,
just,
gop,
that,
image,
mr

<H2>Task 5: Collaborative Filtering
</H2>

<P>In this task, you will use collaborative filtering to predict how a person might rate a business. Ratings are numbers between 1 and 5 corresponding, for example, to how many stars a person gives the business in a review. The data come from <a href="https://www.yelp.com/dataset">Yelp</a>.</P>

<P>To begin, read in the <code>training.csv</code> file containing <em>training</em> data. Unlike many other <code>CSV</code> files that we have worked with previously, the data are not in matrix form (rows and columns) here, but rather a list. Each line in the file contains three comma delimited values: the first corresponds to the ID of a person (user), the second correspond to the ID of a business (item), and the third corresponds to the rating (1 through 5) that the person gave the business in a review. The file contains over 80,000 lines, i.e., over 80,000 ratings that people gave to businesses. Write code below to create a user-item matrix, as described in class, from the data in the file. For each line in the file, you can think of the user (person) ID as indicating the row in the user-item matrix, the item (business) ID as indicating the column in the user-item matrix, and the rating as the value to be inserted in the user-item matrix at the specified row and column. Note that your user-item matrix will have many more than 80,000 entries even though the data contain only ~80,000 ratings. Thus, the matrix will be sparse, i.e., most values in the matrix should correspond to zero since each person rates only a small subset of all businesses.</P>

In [97]:
# Read in training data and create user-item matrix

df = pd.read_csv('training.csv', header=None)
rows = df[0]
cols = df[1]
ratings = df[2]
df.columns = ['user','item','rating']
df1 = df.pivot(index='user', columns='item', values='rating')

df1







item  0     1     2     3     4     5     6     7     8     9     ...  4207  \
user                                                              ...         
0      4.0   4.0   2.0   4.0   2.0   5.0   NaN   NaN   5.0   2.0  ...   NaN   
1      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
2      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
3      NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN   NaN   NaN  ...   NaN   
4      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
913    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
914    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
915    5.0   3.0   3.0   NaN   NaN   5.0   1.0   NaN   NaN   5.0  ...   NaN   
916    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
917    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   

item  4208  4209  4210  4211  4212  4213  4214  4215  4216  
user                                                        
0      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
3      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
913    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
914    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
915    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
916    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
917    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[918 rows x 4217 columns]

<P>Ultimately, our goal is to use filled-in (i.e., non-zero) values in the user-item matrix to predict empty (i.e., zero) values in the user-item matrix. In other words, we want to predict how a person might rate a business, based on how a person has rated other businesses or how a business has been rated by other people. Collaborative filtering.</P>

<P>Below, write a function named <code>prediction</code> that predicts a value (i.e., rating) for a given location (row index and column index) in the user-item matrix. Your prediction algorithm should proceed as follows:<BR>
<UL>
<li>Determine all rows in the matrix that contain non-zero values in the specified column. This will give us information on all people that rated the business of interest.</li>
<li>Using the <code>sklearn</code> unsupervised <code><a href="http://scikit-learn.org/stable/modules/neighbors.html">NearestNeighbors</a></code> algorithm, for the specified row in the user-item matrix, compute the <em>k</em> nearest neighbors among the rows that contain non-zero values in the specified column. This will give us the <em>k</em> most similar people that rated the business of interest. You should use the <code>cosine</code> metric in the <code>NearestNeighbors</code> algorithm for computing distance between rows in a matrix. If there are fewer than <em>k</em> people who have rated the business, you can use everyone who rated the business rather than the <em>k</em> most similar people (since there aren't <em>k</em> such people).</li>
<li>For the <em>k</em> nearest neighbors of the specified row, compute and return the mean value for the specified column. This will give us the average rating of the business from the <em>k</em> most similar people to the person of interest. In other words, we are predicting how someone will rate a business based on the ratings of similar people who have already rated the business.</li>
</UL><BR>
Your <code>prediction</code> function should have at least four inputs: your user-item matrix, a paramter <em>k</em>, a specified row, and a specified column (as well as any other input parameters that you deem helpful). 

In [98]:
# The *prediction* function takes in four (or more) parameters corresponding to
# the user-item matrix, k, the row (person) index, and the column (business) index.
# The function returns the average (mean) rating of a business from the k most
# similar people to the specified person.
import numpy as np
from sklearn.neighbors import NearestNeighbors

def prediction(matrix, k, row, column):
    item = matrix[:][column] 
    xrow = matrix.iloc[row].fillna(0)
    xrow = xrow.to_numpy()
    item_row = np.argwhere(item.notnull().values).tolist()
    user_row = [] 
    for item in item_row:
        for i in item:
            user_row.append(i)
    df2 = matrix.iloc[user_row]
    df2 = df2.fillna(0)
    if df2.shape[0] >= k:
        model = NearestNeighbors(metric='cosine', n_neighbors=k).fit(df2)
    if df2.shape[0] < k:
        model = NearestNeighbors(metric='cosine', n_neighbors=df2.shape[0]).fit(df2)
    dist, indices = model.kneighbors(xrow.reshape(1,-1))
    averages = []
    for row in indices[0]:
        averages.append(df2.iloc[row][column])
    mean = np.mean(averages)
    return(mean)






<P>Let's evaluate how well we do at predicting people's ratings of businesses. The file <code>testing.csv</code> contains approximately 4,000 ratings that different people have given to different businesses. These <em>testing</em> ratings correspond to ~4,000 entries in the user-item matrix that contain zero values. So we can use the user-item matrix (our <em>training</em> data) to predict ratings in these ~4,000 cases and compare our predictions to the actual rating found in the <em>testing</em> data in order to see how accurate our predicted ratings are.</P>

<P>Write code below that reads in <em>testing</em> data from <code>testing.csv</code>. Each line in the file contains three comma delimited values: the first corresponds to the ID of a person (user), the second correspond to the ID of a business (item), and the third corresponds to the rating (1 through 5) that the person gave the business in a review. For each line, use the <code>prediction</code> function that you wrote above to predict the rating that the specified user might give the specified business. Keep track of this predicted rating as well as the actual rating from the <em>testing</em> data. After processing all lines in the <em>testing</em> file, you should have ~4,000 predicted ratings and the same number of actual ratings. To compare how far our predictions are from the actual ratings, we'll compute the root mean-squared error (RMSE) between the two, e.g., by computing the square root of the <a href="http://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html">mean-squared error</a> (MSE). As a point of reference, the instructor found a RMSE a bit under 1.0 depending on the value of <em>k</em> used, i.e., the predicted ratings were a bit less than 1 "star" from the actual ratings on average. Try computing the RMSE on the <em>testing</em> data for each of the following eleven values for the parameter <em>k</em>: 1, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50.</P>

In [99]:
# Read in testing data and for each line in the file,
# calculate the predicted rating of the specified person and business.
# Compare the predicted ratings with the actual ratings in the testing data
# by computing the root mean-squared error (RMSE).
# Compute the RMSE for 11 different values of parameter k: 1,5,10,15,20,25,30,35,40,45,50

df = pd.read_csv('testing.csv', header=None)
df.columns = ['user','item','rating']
results = df[['rating']]

df = df.drop(columns=['rating'])
df.head()
ratings = np.zeros(df.shape[0])
for i in range(df.shape[0]):
    ratings[i] = prediction(df1,10,df.iloc[i][0],df.iloc[i][1])
print(ratings) 

short_list = [] 
for item in results.values:
    for i in item:
        short_list.append(i)
short_array = np.asarray(short_list)
print(short_array)





[4.1 3.5 3.9 ... 4.1 3.9 3.9]
[5 3 4 ... 5 4 4]


In [100]:
from sklearn.metrics import mean_squared_error
y_true = short_array
y_pred = ratings
mean_squared_error(y_true, y_pred)

0.8743359918901499

In [101]:
k = [1, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50]
ratings_k = np.zeros((len(k),df.shape[0]))
rmse = np.zeros(len(k))
for z in range(len(k)):
    for i in range(df.shape[0]):
        ratings_k[z][i] = prediction(df1,k[z],df.iloc[i][0],df.iloc[i][1])
    rmse[z] = mean_squared_error(y_true, ratings_k[z])

In [102]:
rmse

array([1.44915062, 0.94519466, 0.87433599, 0.86201803, 0.85863899,
       0.85894258, 0.8591207 , 0.8597084 , 0.8596854 , 0.85918819,
       0.86063692])

<P><font color="maroon"><u>For which of the eleven values of <em>k</em> do you observe the smallest RMSE? What is the RMSE at this optimal value of <em>k</em>?<u></font></P>

the smallest RMSE value is 0.85863899, which is observed for the k value of 20. 

<P>Above, you predicted how a user might rate a business by seeing how similar users rated the same business (and taking the average value of these similar users' ratings of the business). As an alternative method for predicting the rating a user might give to a business, we could look at similar <em>businesses</em> rather than similar <em>users</em>. In this alternative approach, we will take our user and find similar <em>businesses</em> that they have rated. Our predicted rating will be the mean value of the user's ratings of <em>k</em> similar businesses. Below, write code to compute the RMSE between predicted ratings and actual ratings using eleven different values for the parameter <em>k</em>: 1, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50. Predicted ratings should be determined by calculating, among businesses rated by a specified user, the average rating that the user gave to the <em>k</em> most similar businesses. <em>Hint: You needn't change your <code>prediction</code> function above. Changing a single line of code elsewhere may be sufficient.</em></P>

In [131]:
# Read in testing data and for each line in the file,
# calculate the predicted rating of the specified person and business.
# Compare the predicted ratings with the actual ratings in the testing data
# by computing the root mean-squared error (RMSE).
# Compute the RMSE for 11 different values of parameter k: 1,5,10,15,20,25,30,35,40,45,50


df2 = pd.read_csv('training.csv', header=None)
df2.columns = ['user','item','rating']
df3 = df2.pivot(index='item', columns='user', values='rating')
df3




user  0    1    2    3    4    5    6    7    8    9    ...  908  909  910  \
item                                                    ...                  
0     4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1     4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  5.0   
2     2.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
3     4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  3.0   
4     2.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
4212  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
4213  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
4214  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
4215  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
4216  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   

user  911  912  913  914  915  916  917  
item                                     
0     NaN  NaN  NaN  NaN  5.0  NaN  NaN  
1     NaN  NaN  NaN  NaN  3.0  NaN  NaN  
2     NaN  NaN  NaN  NaN  3.0  NaN  NaN  
3     NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4     NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...   ...  ...  ...  ...  ...  ...  ...  
4212  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4213  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4214  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4215  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4216  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[4217 rows x 918 columns]

In [132]:
df4 = pd.read_csv('testing.csv', header=None)
df4.columns = ['user','item','rating']
df4.columns = ['user','item','rating']

results1 = df4[['rating']]

df4 = df4.drop(columns=['rating'])
df4







user  item
0      562   308
1       58  1731
2      281   427
3      631   141
4      867  3789
...    ...   ...
4410   430   334
4411    62   260
4412   625   758
4413   384  1017
4414   603  1174

[4415 rows x 2 columns]

In [133]:
ratings = np.zeros(df4.shape[0])
for i in range(df4.shape[0]):
    ratings[i] = prediction(df3,10,df4.iloc[i][1],df4.iloc[i][0])
ratings


array([3.9, 3.5, 3.7, ..., 2.9, 3.5, 3.5])

In [134]:
short_list1 = [] 
for item in results1.values:
    for i in item:
        short_list1.append(i)
short_array1 = np.asarray(short_list1)
print(short_array1)

[5 3 4 ... 5 4 4]


In [139]:
k1 = [1, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50]
ratings_k1 = np.zeros((len(k1),df4.shape[0]))
rmse1 = np.zeros(len(k1))
for z in range(len(k1)):
    for i in range(df4.shape[0]):
        ratings_k1[z][i] = prediction(df3,k1[z],df4.iloc[i][1],df4.iloc[i][0])
    rmse1[z] = mean_squared_error(short_array1, ratings_k1[z])

In [140]:
rmse1

array([1.59479049, 0.95605889, 0.88927366, 0.86871762, 0.85748533,
       0.85316512, 0.84970183, 0.84752743, 0.84711902, 0.84662163,
       0.84849161])

<P><font color="maroon"><u>For which of the eleven values of <em>k</em> do you observe the smallest RMSE? What is the RMSE at this optimal value of <em>k</em>?<u></font></P>

The smallest RMSE value is 0.84662163 which is observed for k= 45. 

<P><font color="maroon"><u>For their optimal values of parameter <em>k</em>, which approach led to more accurate predictions: (1) among users that have rated a business, finding the average rating that the <em>k</em> most similar users gave the business or (2) among businesses rated by a user, finding the average rating that the user gave to the <em>k</em> most similar businesses?<u></font></P>

The second approach led to more accurate predictions (smaller rmse values) 

<H2>Submitting your work
</H2>

<P><font color="maroon"><u>Please indicate your name and the names of any partner that worked with you on this project:</u></font></P>

Name(s): Zeynep Yalcin

<P><font color="maroon"><u>Please indicate anyone else that you collaborated with in the process of doing the project:</u></font></P>

Collaborators: went to TA office hours 

<P><font color="maroon"><u>When working on this project, approximately how many hours did you spend on each of (1) Task 1, (2) Task 2, (3) Task 3, (4) Task 4, (5) Task 5, and (6) Total?</u></font></P>

Hours on Task 1: 30 minutes
Hours on Task 2: 1 hour
Hours on Task 3: 45 minutes 
Hours on Task 4: 1 hour
Hours on Task 5: 7.15 hours 
Total hours: 10.5 hours

<P><font color="maroon"><u>When working on this project, did you abide by the <a href="https://www.wellesley.edu/studentlife/aboutus/honor">Honor Code</a> and is all of the work that you are submitting your own and/or your partner's?</u></font></P>

Abide by Honor Code: Yes

<P><font color="maroon"><u>To submit this project, please upload your <code>Project4.ipynb</code> file to the <code>Project4</code> folder that the instructor created and shared with you in your Google drive.</u></font></P>